### PR TITLE
checkpoint also QAT for resuming

### LINKED ISSUE #6457: Can't load_from_checkpoint with QuantizationAwareTraining callback during training
for PL v1.2.0

```py
LightningModel.load_from_checkpoint(str(ckpt_path))
```
gives
```
  File "//checkpoint/utils.py", line 11, in load_pl_model
    pl_model: LightningModel = LightningModel.load_from_checkpoint(str(ckpt_path))
  File "/.venv/lib/python3.8/site-packages/pytorch_lightning/core/saving.py", line 156, in load_from_checkpoint
    model = cls._load_model_state(checkpoint, strict=strict, **kwargs)
  File "//.venv/lib/python3.8/site-packages/pytorch_lightning/core/saving.py", line 204, in _load_model_state
    model.load_state_dict(checkpoint['state_dict'], strict=strict)
  File "//.venv/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1223, in load_state_dict
    raise RuntimeError('Error(s) in loading state_dict for {}:\n\t{}'.format(
RuntimeError: Error(s) in loading state_dict for LightningModel:
Unexpected key(s) in state_dict: "quant.activation_post_process.fake_quant_enabled", "quant.activation_post_process.observer_enabled", "quant.activation_post_process.scale", "quant.activation_post_process.zero_point", "quant.activation_post_process.activation_post_process.eps", "quant.activation_post_process.activation_post_process.min_val", "quant.activation_post_process.activation_post_process.max_val", "model.encoder._bn0.activation_post_process.fake_quant_enabled", "model.encoder._bn0.activation_post_process.observer_enabled", "model.encoder._bn0.activation_post_process.scale", "model.encoder._bn0.activation_post_process.zero_point", .............
```

### PR DESCRIPTION
**Pull request by Borda**
## What does this PR do?

Fixes #6457

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)
- [ ] Did you list all the breaking changes introduced by this pull request?

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py [hunk 1]
```diff
@@ -168,9 +168,6 @@ def restore_training_state(self) -> None:
 
     def restore_callbacks(self) -> None:
         """ Restores all callbacks from the pre-loaded checkpoint. """
-        if not self._loaded_checkpoint:
```
Comment (tchaton) [inline pytorch_lightning/trainer/connectors/checkpoint_connector.py:171]: Why did you remove this ?

File: pytorch_lightning/callbacks/quantization.py [hunk 2]
```diff
@@ -82,6 +83,64 @@ def _recursive_hasattr(obj: Any, attribs: str, state: bool = True) -> bool:
...
+    r"""
+    Quantization allows speeding up inference and decreasing memory requirements
+    by performing computations and storing tensors at lower bitwidths
+    (such as INT8 or FLOAT16) than floating point precision.
+    We use native PyTorch API so for more information
+    see `Quantization <https://pytorch.org/docs/stable/quantization.html#quantization-aware-training>`_.
+
+    .. warning:: ``QuantizationAwareTraining`` is in beta and subject to change.
+
+    The model set for quantization can appear in one of this stages::
```
Comment (awaelchli) [inline pytorch_lightning/callbacks/quantization.py:95]:
```suggestion
    The model set for quantization can appear in one of these stages::
```

File: pytorch_lightning/callbacks/quantization.py [hunk 3]
```diff
@@ -82,6 +83,64 @@ def _recursive_hasattr(obj: Any, attribs: str, state: bool = True) -> bool:
...
+                    Trainer --/                 |
+            ( resume_from_checkpoint )          v
+                       ^                  QAT checkpoints
+                       |----------------------/
+
+    The model enters the process as "vanilla model" and it is prepared for QAT training in ``on_fit_start`` hook.
+    Note that any saved checkpoint includes already collected stat fro performing Quantization conversion,
+     but not any already quantized and/or fused modules/layers.
+    The quantization is performed on the ``on_fit_end`` and so it need sto he saved extra after finished training.
+    If a user wants to continue any past training we encourage to create a Trainer with ``resume_from_checkpoint``.
```
Comment (awaelchli) [inline pytorch_lightning/callbacks/quantization.py:111]:
```suggestion
    The model enters the process as "vanilla model" and it is prepared for QAT training in the ``on_fit_start`` hook.
    Note that any saved checkpoint includes already collected stats for performing Quantization conversion,
    but not any already quantized and/or fused modules/layers.
    The quantization is performed in the ``on_fit_end`` hook and so it needs to be saved extra after training finished.
    If a user wants to continue any past training we encourage to create a Trainer with ``resume_from_checkpoint``.
```

@Borda what do you mean by "so it needs to be saved extra"?

File: pytorch_lightning/callbacks/quantization.py [hunk 4]
```diff
@@ -82,6 +83,64 @@ def _recursive_hasattr(obj: Any, attribs: str, state: bool = True) -> bool:
...
+
+            - 'fbgemm' for server inference.
+            - 'qnnpack' for mobile inference.
+            -  a custom `torch.quantization.QConfig <https://pytorch.org/docs/stable/torch.quantization.html#torch.quantization.QConfig>`_.
+
+        observer_type: allows switching between ``MovingAverageMinMaxObserver`` as "average" (default)
+            and ``HistogramObserver`` as "histogram" which is more computationally expensive.
+        collect_quantization: count or custom function to collect quantization statistics:
+
+            - ``None`` (deafult). The quantization observer is called in each module forward
```
Comment (awaelchli) [inline pytorch_lightning/callbacks/quantization.py:124]:
```suggestion
            - ``None`` (default). The quantization observer is called in each module forward
```

File: pytorch_lightning/callbacks/quantization.py [hunk 5]
```diff
@@ -82,6 +83,64 @@ def _recursive_hasattr(obj: Any, attribs: str, state: bool = True) -> bool:
...
+                (useful for collecting extended statistic when useing image/data augmentation).
```
Comment (awaelchli) [inline pytorch_lightning/callbacks/quantization.py:125]:
```suggestion
                (useful for collecting extended statistics when using image/data augmentation).
```

File: pytorch_lightning/callbacks/quantization.py [hunk 6]
```diff
@@ -82,6 +83,64 @@ def _recursive_hasattr(obj: Any, attribs: str, state: bool = True) -> bool:
...
+                .. code-block:: python
+
+                    def custom_trigger_last(trainer):
+                        return trainer.current_epoch == (trainer.max_epochs - 1)
+
+                    QuantizationAwareTraining(collect_quantization=custom_trigger_last)
+
+        modules_to_fuse: allows you fuse a few layers together as shown in
+            `diagram <https://pytorch.org/docs/stable/quantization.html#quantization-aware-training>`_
+            to find which layer types can be fused, check https://github.com/pytorch/pytorch/pull/43286.
```
Comment (awaelchli) [inline pytorch_lightning/callbacks/quantization.py:139]:
```suggestion
        modules_to_fuse: allows you to fuse a few layers together as shown in
            `diagram <https://pytorch.org/docs/stable/quantization.html#quantization-aware-training>`_.
            To find which layer types can be fused, check https://github.com/pytorch/pytorch/pull/43286.
```

File: pytorch_lightning/callbacks/quantization.py [hunk 7]
```diff
@@ -82,6 +83,64 @@ def _recursive_hasattr(obj: Any, attribs: str, state: bool = True) -> bool:
...
+        input_compatible: preserve quant/dequant layers. This allows to feat any input as to the original model,
+            but break compatibility to torchscript.
```
Comment (awaelchli) [inline pytorch_lightning/callbacks/quantization.py:141]:
```suggestion
        input_compatible: preserve quant/dequant layers. This allows to feed any input as to the original model,
            but breaks compatibility with torchscript.
```

File: pytorch_lightning/callbacks/quantization.py [hunk 8]
```diff
@@ -82,6 +83,64 @@ def _recursive_hasattr(obj: Any, attribs: str, state: bool = True) -> bool:
...
+
+    """  # noa: E501
```
Comment (awaelchli) [inline pytorch_lightning/callbacks/quantization.py:143]: typo here? not sure

```suggestion
    """
```

File: pytorch_lightning/trainer/trainer.py [hunk 9]
```diff
@@ -853,11 +853,11 @@ def _run(self, model: 'pl.LightningModule') -> Optional[Union[_EVALUATE_OUTPUT,
         self.accelerator.setup_environment()
         self._call_setup_hook(model)  # allow user to setup lightning_module in accelerator environment
 
+        # restore callback states
+        self.checkpoint_connector.restore_callbacks()
         # restore modules after setup
         self.checkpoint_connector.restore_datamodule()
         self.checkpoint_connector.restore_model()
-        # restore callback states
-        self.checkpoint_connector.restore_callbacks()
```
Comment (awaelchli) [inline pytorch_lightning/trainer/trainer.py:861]: is it necessary? this will affect all callbacks. maybe it is possible to implement the QAT callback using the setup() method which runs right before restore. What do you think?

Comment (Borda) [reply pytorch_lightning/trainer/trainer.py:861]: > maybe it is possible to implement the QAT callback using the setup() method which runs right before restore

which setup, setup of a model? but the model does not know anything about callback which is added for training
if you write a model which counts on the run only with QAT it can work but how about models which are kind of vanilla and later you just want to train with QAT?

as we spoke earlier, the actual calling `on_load_checkpoint` after the model is restored is not logical and it shall be before... so what I need to have an entry with access to checkpoint before te model is restored and it shall be independent on the model code :rabbit:

Comment (awaelchli) [reply pytorch_lightning/trainer/trainer.py:861]: then I would like to suggest adding a changelog entry under "changed" because we won't be able to make this backward compatible. 

To the best of my knowledge, the order was always

```
DataModule.on_load_checkpoint
Model.on_load_checkpoint
Callbacks.on_load_checkpoint
```

Comment (Borda) [reply pytorch_lightning/trainer/trainer.py:861]: > To the best of my knowledge, the order was always

but in such a case when a callback can be restored and do some model changes before restoring?
On the other hand, the Trainer itself is restored before, so it is:
```
Trainer.restore...
# Callbacks.on_load_checkpoint # suggestions
DataModule.on_load_checkpoint
Model.on_load_checkpoint
Callbacks.on_load_checkpoint # actual now
```

File: pytorch_lightning/callbacks/quantization.py [hunk 10]
```diff
@@ -82,6 +83,64 @@ def _recursive_hasattr(obj: Any, attribs: str, state: bool = True) -> bool:
...
+                    Trainer --/                 |
+            ( resume_from_checkpoint )          v
+                       ^                  QAT checkpoints
+                       |----------------------/
+
+    The model enters the process as "vanilla model" and it is prepared for QAT training in the ``on_fit_start`` hook.
+    Note that any saved checkpoint includes already collected stats for performing Quantization conversion,
+    but not any already quantized and/or fused modules/layers.
+    The quantization is performed in the ``on_fit_end`` hook and so it needs to be saved extra after training finished.
+    If a user wants to continue any past training we encourage to create a Trainer with ``resume_from_checkpoint``.
```
Comment (Borda) [inline pytorch_lightning/callbacks/quantization.py:111]:
```suggestion
    The model enters the process as a "vanilla model" and it is prepared for QAT training in the ``on_fit_start`` hook.
    Note that any saved checkpoint includes already collected stats for performing Quantization conversion,
    but not any already quantized and/or fused modules/layers.
    The quantization is performed in the ``on_fit_end`` hook and so the model needs to be saved independently after the training is finished.
    If a user wants to continue any past training we encourage to create a Trainer with ``resume_from_checkpoint``.
```

File: pytorch_lightning/trainer/trainer.py (hunk 9 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/trainer.py:861]: > On the other hand, the Trainer itself is restored before, so it is:

no, this is not the case. The trainer, in terms of loop state, optimizers etc. gets restored last, just before training begins. 

the `on_load_checkpoint` hooks are the very first hooks that run while restoring so the user has a chance to modify the checkpoint dict.

File: pytorch_lightning/callbacks/quantization.py [hunk 11]
```diff
@@ -91,45 +92,58 @@ class QuantizationAwareTraining(Callback):
...
-    Args:
+        TRAINER TRANSITIONS
+                          ( on_fit_start )               ( on_fit_end )
+        MODEL STATES
+            vanilla model      --->       QuantAwareTrain     --->     quantized model
+                               /                |
+                    Trainer --/                 |
+            ( resume_from_checkpoint )          v
+                       ^                  QAT checkpoints
+                       |----------------------/
```
Comment (carmocca) [inline pytorch_lightning/callbacks/quantization.py:105]: This doesn't look very easy to read at all...

File: pytorch_lightning/trainer/trainer.py (hunk 9 above)
Comment (carmocca) [reply pytorch_lightning/trainer/trainer.py:861]: I would be very careful about changing the order.

In master, callbacks can expect the `LightningModule` to be fully loaded before loading themselves, one example is the `BaseFinetuningCallback`.

https://github.com/PyTorchLightning/pytorch-lightning/blob/b7dbcc3e13a2e34e9fc1850d461909caf8ab11e0/pytorch_lightning/callbacks/finetuning.py#L95-L103

Changing this could break many existing callbacks.

I am not very familiar with what you are trying to fix. Can you elaborate on why do you need to `self.prepare_model(pl_module)` before the model is loaded?

Comment (Borda) [reply pytorch_lightning/trainer/trainer.py:861]: @tchaton is fixing finetuning.py

Comment (Borda) [reply pytorch_lightning/trainer/trainer.py:861]: > Can you elaborate on why do you need to `self.prepare_model(pl_module)` before the model is loaded?

I need to add some states to the modules which are used for collecting stats, otherwise, the loading fails as checkpointed model contains attributes that are not presented in the vanilla model

File: pytorch_lightning/callbacks/quantization.py (hunk 11 above)
Comment (Borda) [reply pytorch_lightning/callbacks/quantization.py:105]: any other suggestions?

Comment (carmocca) [reply pytorch_lightning/callbacks/quantization.py:105]: My suggestion is don't try to draw a graph in plaintext 😂 

Rewrite it in pseudocode

File: pytorch_lightning/trainer/trainer.py (hunk 9 above)
Comment (carmocca) [reply pytorch_lightning/trainer/trainer.py:861]: But why don't you remove any quantization related data on the model before the model gets checkpointed?

Then loading it wouldn't fail without the callback

Any of the callback logic should be entirely saved by the callback, and when loaded, re-applied to the model.

File: pytorch_lightning/callbacks/quantization.py (hunk 11 above)
Comment (Borda) [reply pytorch_lightning/callbacks/quantization.py:105]: I would rather have some chart the pseudocode...
@t-vi @aribornstein what do you think?

File: pytorch_lightning/trainer/trainer.py (hunk 9 above)
Comment (Borda) [reply pytorch_lightning/trainer/trainer.py:861]: as poke with @t-vi of we drop all related attributes that resume does not make sense as you removed them...

> Any of the callback logic should be entirely saved by the callback, and when loaded, re-applied to the model.

so I will save the full model again as a callback state, that is what you suggest?

**Comment by Borda** [comment]
@ananthsub what is your suggestion on the restoring order - callback vs Model?
@edenafek as there is not clear consensus, shall we move it to 1.5?

**Comment by tchaton** [comment]
Hey @carmocca,

I moved the logic from the BaseFinetuning Callback on `on_fit_start`, so I think it should be fine.

Best,
T.C

File: pytorch_lightning/callbacks/quantization.py (hunk 11 above)
Comment (tchaton) [reply pytorch_lightning/callbacks/quantization.py:105]: Yes, this is un-clear with me either. Missing what are the on_fit_start, on_fit_end consequences on the model.

Comment (Borda) [reply pytorch_lightning/callbacks/quantization.py:105]: on_fit_start, on_fit_end are Trainer hooks so what else do you suggest, otherwise I ll just delete it :rabbit:

File: pytorch_lightning/trainer/callback_hook.py [hunk 12]
```diff
@@ -302,6 +303,11 @@ def on_load_checkpoint(self, checkpoint):
                 UserWarning,
             )
 
+        # in case QAT was used for original training and not inserted void  after resume
+        is_qat = any(isinstance(cb, QuantizationAwareTraining) for cb in self.callbacks)
+        if QuantizationAwareTraining in callback_states and not is_qat:
+            self.callbacks.append(QuantizationAwareTraining())
+
```
Comment (ananthsub) [inline pytorch_lightning/trainer/callback_hook.py:310]: i am very opposed to special casing callbacks like this

Comment (tchaton) [reply pytorch_lightning/trainer/callback_hook.py:310]: Would you prefer to raise a MisConfigurationError instead ? From my understanding, this is the only other alternative.

Comment (Borda) [reply pytorch_lightning/trainer/callback_hook.py:310]: ok, I'll change it as misconfiguration :rabbit:

**Comment by Borda** [comment]
@ananthsub what is your suggestion on the restoring order - callback vs Model?

File: pytorch_lightning/trainer/callback_hook.py (hunk 12 above)
Comment (tchaton) [reply pytorch_lightning/trainer/callback_hook.py:310]: @Borda  Mind raising a MisConfigurationError ?
